src/interfaces/chain.cpp

```diff
@@ -358,22 +359,39 @@ class ChainImpl : public Chain
     {
         ::uiInterface.ShowProgress(title, progress, resume_possible);
     }
-    std::unique_ptr<Handler> handleNotifications(Notifications& notifications) override
+    std::unique_ptr<Handler> handleNotifications(Notifications& notifications, SnapshotFn snapshot_fn) override
+        LOCKS_EXCLUDED(::cs_main, ::mempool.cs)
     {
-        return MakeUnique<NotificationsHandlerImpl>(*this, notifications);
+        AssertLockNotHeld(::cs_main);
+        AssertLockNotHeld(::mempool.cs);
+
+        // Declare packaged task to run in notification thread, and to first
+        // send the snapshot before enabling subsequent notifications.
+        std::vector<CTransactionRef> snapshot;
+        std::packaged_task<std::unique_ptr<Handler>()> task([&] {
+            if (snapshot_fn) snapshot_fn(std::move(snapshot));
+            return MakeUnique<NotificationsHandlerImpl>(*this, notifications);
+        });
+
+        // Snapshot mempool transactions, then schedule the task to execute.
+        // Hold locks while scheduling the task so notifications about added and
+        // removed transactions after the snapshot arrive after the snapshot.
```

Comment:
also "arrive after task() is called" right?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r271410636

> also "arrive after task() is called" right?

Yes task() is what sends the snapshot, so that's what this is referring to. Replaced "transactions after the snapshot arrive after the snapshot" with "transactions after taking the snapshot arrive after the snapshot callback in task()" to reference it and be more specific.